sse_client/event.go

```diff
@@ -0,0 +1,119 @@
+package sse_client
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"io"
+	"strconv"
+)
+
+type Event struct {
+	ID    string
+	Event string
+	//The data field for the message. When the EventSource receives multiple consecutive lines that begin with data:,
+	//it will concatenate them, inserting a newline character between each one.
+	//Trailing newlines are removed.
+	Data []byte
+	//The reconnection time to use when attempting to send the event.
+	//This must be an integer, specifying the reconnection time in milliseconds.
+	//If a non-integer value is specified the field is ignored.
+	ReconnectDelay uint64
+}
+
+func parseEvent(reader *bufio.Reader) (Event, error) {
+	e := Event{}
+	for dispatch := false; !dispatch; {
+		//TODO: Use scanner use ReadLine
+		line, err := reader.ReadBytes('\n')
+		if err == io.EOF {
+			dispatch = e.parseLine(line)
+			if !dispatch {
+				return e, errors.New("Unexpected EOF")
+			}
+			return e, io.EOF
+		}
+		if err != nil {
+			return e, err
+		}
+		dispatch = e.parseLine(line)
+	}
+	return e, nil
+}
+
+// parseLine is implementation of event stream interpretation w3c standard. Is designed to be run multiple time.
+// `line` argument should be line readed from buffer without new line character.
+// Returns true when event is ready for dispatch
+// See: https://www.w3.org/TR/eventsource/#event-stream-interpretation
+// Reference implementation: https://github.com/WebKit/webkit/blob/9f191f/Source/WebCore/page/EventSource.cpp#L272-L363
+func (e *Event) parseLine(line []byte) bool {
+
+	line = bytes.TrimSuffix(line, []byte{'\n'})
+	line = bytes.TrimSuffix(line, []byte{'\r'})
+
+	//println(string(line))
+
+	//If the line is empty (a blank line)
+	if len(line) == 0 {
+		//Dispatch the event, as defined below.
+		return !e.isEmpty()
+	}
+
+	//If the line starts with a U+003A COLON character (:)
+	if bytes.HasPrefix(line, []byte{':'}) {
```

Comment:
instead of allocating it every time introduce it as var. same applies to every other occurance